src/main/scala/dpla/ingestion3/harvesters/oai/OaiResponseProcessor.scala

```diff
@@ -60,20 +80,102 @@ object OaiResponseProcessor {
     * @return Seq[String]
     *         The set ids.
     */
-  def getSetIdsFromRecord(xml: NodeSeq): Seq[String] = {
+  def getSetIdsFromRecord(xml: Node): Seq[String] = {
     val sets = xml \\ "setSpec"
     sets.map(s => s.text)
   }
 
   /**
-    * Get the error property if it exists
+    * Parses sets from an OaiSource, and filter if necessary according to a
+    * whitelist or blacklist.
     *
-    * @return Option[String]
-    *         The error code if it exists otherwise empty string
+    * @param source OaiSource
+    *               The single-page response to a single OAI query
+    *
+    * @param setFilter: PartialFunction[(Seq[OaiSet]), Seq[OaiSet]])
+    *                   A partial function indicating how sets should be filtered,
+    *                   ie. according to a whitelist or blacklist.
+    */
+  def getSets(source: OaiSource, setFilter: PartialFunction[(Seq[OaiSet]), Seq[OaiSet]]): OaiResponse = {
+
+    // Parse all sets from XML.
+    def getSetsFromXml(xml: Node): Seq[OaiSet] = {
+      val xmlSets = xml \\ "OAI-PMH" \\ "set"
+
+      xmlSets.flatMap(set => set.headOption match {
```

Comment:
This might be cleaner with a `for`, too, and you wouldn't need the pattern match. Winging it in github:

`for (set <- xml \\ "OAI-PMH" \\ "set") yield Some(OaiSet(getSetIdentifier(set), node.text, source))`

Also, if these nodes are definitely children of one another, using `\` is preferred because `\\` means you want it to dig through all the descendants, not just the first level.